### ISSUE TITLE
Pointer aliasing with unused parameter

### ISSUE THREAD

**Issue by hasali19** [gfx-rs/wgpu#4397]
The spec says that aliasing is invalid if the aliases access the same memory location and at least one of the accesses is a write.

```
@group(0)
@binding(1)
var<storage, read_write> s_out: i32;

var<private> flag: bool;

fn func(p: ptr<private, bool>) {
    flag = true;
}

@compute
@workgroup_size(1)
fn main() {
    func(&flag);
    if (flag) {
        s_out = 1;
    } else {
        s_out = 2;
    }
}
```

In this code, a pointer to `flag` is passed to `func` but it is never used, and `flag` is only ever accessed through the originating variable identifier. So based on my understanding (or have I misunderstood the spec?) this should be fine. However, on DirectX this behaves unexpectedly with an output value of `2`. On Vulkan it works as expected, outputting `1`.

Is this a bug or is it forbidden even if the memory location isn't accessed through the pointer?

**Comment by jimblandy** [comment]
I believe this code is now forbidden by WGSL, per [the restrictions in the section Aliasing](https://gpuweb.github.io/gpuweb/wgsl/#aliasing), and the bug is that Naga fails to validate it out.

**Comment by bavalpey** [comment]
Correct me if I am wrong, but it seems that Naga still does not fully implement the Alias Analysis as required by the WGSL spec (noted in the comment above).

Doesn't this mean that wgpu, which uses Naga for validation, is not conformant with the WGSL spec?

**Comment by ErichDonGubler** [comment]
@bavalpey:

> Doesn't this mean that wgpu, which uses Naga for validation, is not conformant with the WGSL spec?

Yep! And we're trying to fix it. WebGPU compliance is a major driver for Firefox contributors to WGPU, and you can follow along at <https://github.com/orgs/gfx-rs/projects/7/views/5> if you'd like to see the priorities we've identified.

**Comment by bavalpey** [comment]
Thanks for your response! I do have some stake in this. Sharing thoughts below.

I'm working on an analysis for wgsl and am using Naga to drive it, hooking into it directly after its validation pass. I would like to depend on on the guarantees of Alias Analysis. I suppose I _could_ just continue with the assumption that the aliasing requirements of WGSL aren't violated.

Are there currently plans to implement the alias analysis? I didn't see it explicitly mentioned on that view you shared.

One of the things that would make my life easier is if there was some way, in Naga, to figure out the [originating variable](https://gpuweb.github.io/gpuweb/wgsl/#originating-variable) for the pointer expression. This is done for a different reason in the uniformity analysis, where we see `assignable_global` added in `ExpressionInfo`.

I would presume that, similar to the uniformity requirement, there would be an additional field to expressions in the `expression_info` field in the ModuleInfo returned by the validation layer that provides this information for all pointer types, not just globals.

I do have one question, and perhaps someone here can shed some light on it.  The WGSL spec says that

> Execution of a WGSL function [must not](https://gpuweb.github.io/gpuweb/wgsl/#shader-creation-error) potentially access memory through aliased root identifiers,

However, in the proceeding discussion, the analysis only discusses prohibitions on aliases of parameters. It doesn't discuss aliases to other variables. I am curious, then, as to the legality of something like this:
```wgsl
@group(0) @binding(0) var<storage, read_write> x: i32;

fn foo() -> i32 {
  let a = &x;
  *a = 4;
  return x;
}

**Comment by ErichDonGubler** [comment]
@bavalpey: The issue you're looking for is <https://github.com/gfx-rs/wgpu/issues/4369>.

I don't have anything immediately useful to contribute to the discussion, but @teoxoy or @jimblandy may.

**Comment by bavalpey** [comment]
Not sure how that issue is related. I don't have questions about uniformity analysis.

**Comment by ErichDonGubler** [comment]
@bavalpey: Ah, I must be confusing the two, then. I hadn't read the [`Aliasing` section](https://gpuweb.github.io/gpuweb/wgsl/#aliasing) [mentioned](https://github.com/gfx-rs/wgpu/issues/4397#issuecomment-1856277345) by @jimblandy, and didn't realize that these are two separate analyses.

This is likely something we should add to our board, then! Sorry for the confusion. 😓
